test/integration/controller/core/admissioncheck_controller_test.go

```diff
@@ -0,0 +1,133 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	utiltesting "sigs.k8s.io/kueue/pkg/util/testing"
+	"sigs.k8s.io/kueue/test/util"
+)
+
+// +kubebuilder:docs-gen:collapse=Imports
+
+var _ = ginkgo.Describe("AdmissionCheck controller", func() {
+	var ns *corev1.Namespace
+
+	ginkgo.BeforeEach(func() {
+		ns = &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: "core-admissioncheck-",
+			},
+		}
+		gomega.Expect(k8sClient.Create(ctx, ns)).To(gomega.Succeed())
+	})
+
+	ginkgo.AfterEach(func() {
+		gomega.Expect(util.DeleteNamespace(ctx, k8sClient, ns)).To(gomega.Succeed())
+	})
+
+	ginkgo.When("one clusterQueue references admissionChecks", func() {
+		var admissionCheck *kueue.AdmissionCheck
+		var clusterQueue *kueue.ClusterQueue
+
+		ginkgo.BeforeEach(func() {
+			admissionCheck = utiltesting.MakeAdmissionCheck("check1").Obj()
+			clusterQueue = utiltesting.MakeClusterQueue("foo").
+				AdmissionChecks("check1").
+				Obj()
+
+			gomega.Expect(k8sClient.Create(ctx, admissionCheck)).To(gomega.Succeed())
+			gomega.Expect(k8sClient.Create(ctx, clusterQueue)).To(gomega.Succeed())
+
+			ginkgo.By("Wait for the queue to become active", func() {
+				gomega.Eventually(func() bool {
+					var cq kueue.ClusterQueue
+					err := k8sClient.Get(ctx, client.ObjectKeyFromObject(clusterQueue), &cq)
+					if err != nil {
+						return false
+					}
+					return apimeta.IsStatusConditionTrue(cq.Status.Conditions, kueue.ClusterQueueActive)
+				}, util.Timeout, util.Interval).Should(gomega.BeTrue())
+			})
+		})
+
+		ginkgo.AfterEach(func() {
+			util.ExpectClusterQueueToBeDeleted(ctx, k8sClient, clusterQueue, true)
+			util.ExpectAdmissionCheckToBeDeleted(ctx, k8sClient, admissionCheck, true)
+		})
+
+		ginkgo.It("Should delete the admissionCheck when the corresponding clusterQueue no longer uses the admissionCheck", func() {
+			ginkgo.By("Try to delete admissionCheck")
+			gomega.Expect(util.DeleteAdmissionCheck(ctx, k8sClient, admissionCheck)).To(gomega.Succeed())
+			var ac kueue.AdmissionCheck
+			gomega.Eventually(func() []string {
+				gomega.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(admissionCheck), &ac)).To(gomega.Succeed())
+				return ac.GetFinalizers()
+			}, util.Timeout, util.Interval).Should(gomega.BeComparableTo([]string{kueue.ResourceInUseFinalizerName}))
```

Comment:
```suggestion
			var ac kueue.AdmissionCheck
			gomega.Eventually(func() []string {
				gomega.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(admissionCheck), &ac)).To(gomega.Succeed())
				return ac.GetFinalizers()
			}, util.Timeout, util.Interval).Should(gomega.BeComparableTo([]string{kueue.ResourceInUseFinalizerName}))
			gomega.Expect(util.DeleteAdmissionCheck(ctx, k8sClient, admissionCheck)).To(gomega.Succeed())
```

I think we should check the finalizer before removing the AdmissionCheck.

Reply:
Whether the finalizer is set is an implementation detail. What is important is that an object is not deleted while it's still needed, so I prefer the way the test is currently written.

Reply:
I see. Thanks for the clarifications.